[zh] Sync Better External Authorization blog into Chinese (#13500)
istio/istio.io: content/zh/blog/2021/better-external-authz/index.md

[嵌套 JWT 声明支持](/zh/docs/tasks/security/authorization/authz-jwt/)等等。
这些特性提高了授权策略的灵活性，但仍有许多场景无法通过该模型支持，例如：

- 您拥有自己的内部授权系统，该系统无法轻松迁移或替换至授权策略中。

- 您想与使用 Istio 中[底层 Envoy 配置 API](/zh/docs/reference/config/networking/envoy-filter/)
  （例如 [Open Policy Agent](https://www.openpolicyagent.org/docs/latest/envoy-introduction/)
  或 [`oauth2` 代理](https://github.com/oauth2-proxy/oauth2-proxy)）
  或者根本无法正常工作的第三方解决方案进行集成。

- 授权策略缺少在您场景中所需的语义内容。

## 解决方案 {#solution}

在 Istio 1.9 中，我们通过引入
[`CUSTOM` 操作](/zh/docs/reference/config/security/authorization-policy/#AuthorizationPolicy-Action)实现了授权策略的可扩展性，
它允许您将访问控制决策委托给外部授权服务。

`CUSTOM` 操作允许您将 Istio 与外部授权系统集成，
该系统实现了自己的自定义授权逻辑。下图展示了此集成方式的顶层架构：

{{< image width="100%"
    link="./external_authz.svg"
    caption="外部授权架构"
    >}}

在进行配置时，网格管理员使用 `CUSTOM` 操作对授权策略进行配置，
用于在代理（网关或 Sidecar）上启用外部授权。
管理员应确认外部身份验证服务已启动且正在运行。

在运行时中：

1. 请求被代理拦截，代理将根据用户在授权策略中的配置向外部授权服务发送检查请求。

1. 外部授权服务将决定是否允许请求通过。

1. 如果允许，请求将继续，并将由 `ALLOW`/`DENY` 操作定义的任意本地授权强制执行。

1. 如果被拒绝，请求将立即被终止。

让我们看一下带有 `CUSTOM` 操作的示例授权策略：

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ext-authz
  namespace: istio-system
spec:

>   # 选择器适用于 istio-system 命名空间中的入口网关。

Reviewer: ```suggestion
  # The selector applies to the ingress gateway in the istio-system namespace.
```
Normally, we need not to translate any text in a code block in the istio doc

  selector:
    matchLabels:
      app: istio-ingressgateway
  # “CUSTOM” 操作将访问控制委托给外部授权程序，
  # 这与在代理内部强制执行访问控制的 ALLOW/DENY 操作不同。
  action: CUSTOM
  # provider 指定在 meshconfig 中定义的外部授权者的名称，
  # 它告知我们在哪里以及如何与外部 auth 服务对话。稍后我们将对此进行更多介绍。
  provider:
    name: "my-ext-authz-service"
  # 该规则指定只有当请求路径有前缀“/admin/”时才会触发访问控制。
  # 这使您可以轻松地根据请求本身启用或禁用外部授权，避免在不需要时进行额外的外部检查请求。
  rules:
  - to:
    - operation:
        paths: ["/admin/*"]
{{< /text >}}

它引用了一个名为 `my-ext-authz-service` 的提供程序，
该提供程序被定义在网格配置中：

{{< text yaml >}}
extensionProviders:
# 名称“my-ext-authz-service”在其授权策略的 provider 字段中被引用。
- name: "my-ext-authz-service"
  # “envoyExtAuthzGrpc”字段指定外部授权服务的类型由 Envoy ext-authz filter gRPC API 实现。
  # 另一种被支持的类型是 Envoy ext-authz 过滤器的 HTTP API。
  # 在 https://www.envoyproxy.io/docs/envoy/v1.16.2/intro/arch_overview/security/ext_authz_filter 中查看更多信息。
  envoyExtAuthzGrpc:

>     # service 和 port 字段指定外部认证服务的地址，
>     # “ext-authz.istio-system.svc.cluster.local”表示该服务部署在网格中。
>     # 它也可以是定义为在网格外部的服务，或者是在相同 Pod 内定义的独立容器。

Reviewer: ```suggestion
    # The service and port specifies the address of the external auth service, "ext-authz.istio-system.svc.cluster.local"
    # means the service is deployed in the mesh. It can also be defined out of the mesh or even inside the pod as a separate
    # container.
```
Same as above

1. 外部授权服务将决定是否允许请求通过。

1. 如果允许，请求将继续，并将由 `ALLOW`/`DENY` 操作定义的任意本地授权强制执行。

1. 如果被拒绝，请求将立即被终止。

让我们看一下带有 `CUSTOM` 操作的示例授权策略：

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ext-authz
  namespace: istio-system
spec:
  # The selector applies to the ingress gateway in the istio-system namespace.
  selector:
    matchLabels:
      app: istio-ingressgateway
  # The action "CUSTOM" delegates the access control to an external authorizer, this is different from
  # the ALLOW/DENY action that enforces the access control right inside the proxy.
  action: CUSTOM
  # The provider specifies the name of the external authorizer defined in the meshconfig, which tells where and how to
  # talk to the external auth service. We will cover this more later.
  provider:
    name: "my-ext-authz-service"
  # The rule specifies that the access control is triggered only if the request path has the prefix "/admin/".
  # This allows you to easily enable or disable the external authorization based on the requests, avoiding the external
  # check request if it is not needed.
  rules:
  - to:
    - operation:
        paths: ["/admin/*"]
{{< /text >}}

> 它引用了一个名为 `my-ext-authz-service` 的提供程序，
> 该提供程序被定义在网格配置中：

Reviewer: ```suggestion
此示例引用了一个在网格配置中定义的、名为 `my-ext-authz-service` 的提供程序：
```

- name: "my-ext-authz-service"
  # The "envoyExtAuthzGrpc" field specifies the type of the external authorization service is implemented by the Envoy
  # ext-authz filter gRPC API. The other supported type is the Envoy ext-authz filter HTTP API.
  # See more in https://www.envoyproxy.io/docs/envoy/v1.16.2/intro/arch_overview/security/ext_authz_filter.
  envoyExtAuthzGrpc:
    # The service and port specifies the address of the external auth service, "ext-authz.istio-system.svc.cluster.local"
    # means the service is deployed in the mesh. It can also be defined out of the mesh or even inside the pod as a separate
    # container.
    service: "ext-authz.istio-system.svc.cluster.local"
    port: 9000
{{< /text >}}

授权策略中的 [`CUSTOM` 操作](/zh/docs/reference/config/security/authorization-policy/#AuthorizationPolicy-Action)表示在运行时中启用外部授权，
可以配置为根据请求有条件地触发外部授权，
并且使用您已经用于其他操作的相同规则进行外部授权。

外部授权服务当前在 [`meshconfig` API](/zh/docs/reference/config/istio.mesh.v1alpha1/#MeshConfig-ExtensionProvider)
中定义并通过其名称进行引用。它可以部署在任何使用或不使用代理的网格环境中。
如果使用代理，您可以进一步使用 `PeerAuthentication`
配置在代理和外部授权服务之间开启 mTLS。

`CUSTOM` 操作目前仍然处于**实验阶段**；API

> 可能会基于用户反馈进行针对后续版本的不兼容修改。当授权策略规则与 `CUSTOM`

Reviewer: ```suggestion
可能会基于用户反馈针对后续版本进行不兼容的修改。当授权策略规则与 `CUSTOM`
```

操作一起使用时，其目前不支持身份验证字段（例如源主体或 JWT 声明）。
在单独的工作负载中只允许使用一个提供程序，但您仍然可以在不同的工作负载上使用不同的提供程序。

> 有关详细信息，请参阅[更好的外部授权设计文档](https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/edit#)。

Reviewer: ```suggestion
有关详细信息，请参阅 [Better External Authorization 设计文档](https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/edit#)。
```
点击之后没有中文版，如果翻译了，会让人觉得是不是点错了

## OPA 示例 {#example-with-opa}

在本节中，我们将演示如何使用 `CUSTOM` 操作以及
Open Policy Agent 作为入口网关上的外部授权程序。我们将有条件地在除
`/ip` 之外的所有路径上启用外部授权。

您还可以参考[外部授权任务](/zh/docs/tasks/security/authorization/authz-custom/)来获得使用
`ext-authz` 服务器示例的更基础介绍。

### 创建 OPA 策略示例 {#create-the-example-opa-policy}

运行以下命令创建一个 OPA 策略，如果路径的前缀与 JWT
令牌中的声明“path”（base64 编码）匹配，则允许该请求：

{{< text bash >}}
$ cat > policy.rego <<EOF
package envoy.authz

import input.attributes.request.http as http_request

default allow = false

token = {"valid": valid, "payload": payload} {
    [_, encoded] := split(http_request.headers.authorization, " ")
    [valid, _, payload] := io.jwt.decode_verify(encoded, {"secret": "secret"})
}

allow {
    is_token_valid
    action_allowed
}

is_token_valid {
  token.valid
  now := time.now_ns() / 1000000000
  token.payload.nbf <= now
  now < token.payload.exp
}

action_allowed {
  startswith(http_request.path, base64url.decode(token.payload.path))
}
EOF
$ kubectl create secret generic opa-policy --from-file policy.rego
{{< /text >}}

### 部署 httpbin 和 OPA {#deploy-httpbin-and-opa}

> 启用边车注入：

Reviewer: ```suggestion
启用 Sidecar 注入：
```
